File: fn/src/main/java/org/sqlbuilder/fn/objects/Label.java
package org.sqlbuilder.fn.objects;

import org.sqlbuilder.common.Insertable;
import org.sqlbuilder.common.Utils;
import org.sqlbuilder.common.RequiresIdFrom;
import org.sqlbuilder.fn.types.LabelType;

import java.util.Comparator;
import java.util.HashSet;
import java.util.Set;

public class Label implements Insertable
{
	public static final Set<Label> SET = new HashSet<>();

	private final String name;

	private final Integer itypeid;

	private final int feid;

	private final int start;

	private final int end;

	public final Layer layer;

	@SuppressWarnings("UnusedReturnValue")
	public static Label make(final edu.berkeley.icsi.framenet.LabelType label, final Layer layer)
	{
		var l = new Label(label, layer);
		LabelType.add(l.name);
		SET.add(l);
		return l;
	}

	private Label(final edu.berkeley.icsi.framenet.LabelType label, final Layer layer)
	{
		this.name = label.getName();
		this.itypeid = label.getItype() == null ? null : label.getItype().intValue();
		this.feid = label.getFeID();
		this.start = label.getStart();
		this.end = label.getEnd();
		this.layer = layer;
	}

	public static final Comparator<Label> COMPARATOR = Comparator.comparing(l -> l.name);

	@RequiresIdFrom(type = Layer.class)
	@RequiresIdFrom(type = LabelType.class)
	@Override
	public String dataRow()
	{
		// (labelid),labeltype,labelitypeid,feid,start,end,layerid,fgcolor,bgcolor,cby
		return String.format("%d,%s,%s,%s,%s,%s", //
				LabelType.getIntId(name), //
				Utils.nullableInt(itypeid), //
				Utils.zeroableInt(feid), //
				Utils.zeroableInt(start), //
				Utils.zeroableInt(end), //
				layer.getSqlId());
	}

	@Override
	public String comment()
	{
		return String.format("type=%s", name);
	}

	@Override
	public String toString()
	{
		return String.format("[LAB label=%s layer=%s]", this.name, this.layer);
	}
}

File: fn/src/main/java/org/semantikos/fn/objects/Label.kt
package org.semantikos.fn.objects

import edu.berkeley.icsi.framenet.LabelType
import org.semantikos.annotations.RequiresIdFrom
import org.semantikos.common.Insertable
import org.semantikos.common.Utils.nullableInt
import org.semantikos.common.Utils.zeroableInt
import org.semantikos.fn.types.LabelType.add
import org.semantikos.fn.types.LabelType.getIntId

class Label private constructor(label: LabelType, val layer: Layer) : Insertable, Comparable<Label> {

    private val name: String = label.name

    private val itypeid: Int? = if (label.itype == null) null else label.itype.intValue()

    private val feid: Int = label.feID

    private val start: Int = label.start

    private val end: Int = label.end

    @RequiresIdFrom(type = Layer::class)
    @RequiresIdFrom(type = org.semantikos.fn.types.LabelType::class)
    override fun dataRow(): String {
        return "${getIntId(name)},${nullableInt(itypeid)},${zeroableInt(feid)},${zeroableInt(start)},${zeroableInt(end)},${layer.getSqlId()}"
    }

    // O R D E R

    override fun compareTo(that: Label): Int {
        return COMPARATOR.compare(this, that)
    }

    // I N S E R T

    override fun comment(): String {
        return "type=$name"
    }

    override fun toString(): String {
        return "[LAB label=$name layer=$layer]"
    }

    companion object {

        val COMPARATOR: Comparator<Label> = Comparator
            .comparing<Label, String> { it.name }
            .thenComparing<Layer> { it.layer }
            .thenComparing<Int> { it.start }
            .thenComparing<Int> { it.end }

        val SET = HashSet<Label>()

        fun make(label: LabelType, layer: Layer): Label {
            val l = Label(label, layer)
            add(l.name)
            SET.add(l)
            return l
        }
    }
}
